truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/source/SourceSection.java

```diff
@@ -76,6 +83,30 @@
         this.startColumn = startColumn;
         this.charIndex = charIndex;
         this.charLength = charLength;
+        this.tags = tags;
+        assert tagsAreNonNullAndInterned(tags) : "All tags set for a source section must be interned and non-null.";
+    }
+
+    @SuppressFBWarnings("ES_COMPARING_STRINGS_WITH_EQ")
+    private static boolean tagsAreNonNullAndInterned(String[] tags) {
+        for (int i = 0; i < tags.length; i++) {
+            if (tags[i] == null) {
+                return false;
+            }
+            if (tags[i].intern() != tags[i]) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Returns a set of tags that are associated with this source section as array. The returned
+     * array must be considered read-only. All elements of the array were asserted to be non-null
+     * and interned for faster comparison.
+     */
+    public String[] getTags() {
```

Comment:
Hm, I am not sure I understand. You want this to be only accessible for the framework itself?

I use it to access tags for Json serialization of source sections, and because my tool needs to implement set semantics for tags, since it is not in the API (and because I got very specific requirements on which tags are combinable).

And, I'd rather rewrite my code later on in case this is changed again then having to jump through 1000 layers of API.